Source/WebCore/platform/graphics/ImageBuffer.cpp

```diff
@@ -415,9 +419,15 @@ void ImageBuffer::draw(GraphicsContext& destContext, const FloatRect& destRect,
     srcRectScaled.scale(resolutionScale());
 
     if (auto* backend = ensureBackendCreated()) {
-        if (auto image = copyNativeImageForDrawing(&destContext == &context() ? CopyBackingStore : DontCopyBackingStore))
-            destContext.drawNativeImage(*image, backendSize(), destRect, srcRectScaled, options);
-        backend->finalizeDrawIntoContext(destContext);
+        if (&destContext == &context()) {
```

Comment:
I cannot think of a case where the backend could do arbitrary in-place update of itself.

> Could this logic go inside copyNativeImageForDrawing?

It can once RemoteImageBufferProxy is removed and the function is made non-overridable.

Overrides might miss the case and as such have the bug. E.g. today just moving the implementation to the ImageBuffer::copyNativeImageForDrawing would cause the transient bug that RemoteImageBufferProxy::copyNativeImage(DontCopyBackingStore) would be called for GPUP-only image buffers. Today this would not make a difference but tomorrow, who knows!

I'll land as is and we can develop further in further patches.